File: web/src/api/manage/telemetry.ts
import { http } from '../http'

export interface OverviewResp {
  pv: number
  uv: number
  uniqueUsers: number
  errorCount: number
  apiCount: number
  avgApiMs: number
  windowDays: number
}

export interface IssueResolution {
  id: number
  category: string
  path: string
  label: string
  resolvedAt: string
  resolvedCommitId: string
  resolvedBy: number
  note: string
  reopenedCount: number
}

/** issueStatus:
 *   unresolved = 未处理 (默认 active 列表)
 *   reopened   = 已解决但 24h 灰度期后又触发, 默认 active 列表显示
 *   resolved   = 已解决且 <= 30 天内无 24h 后的新事件, 不在 active 列表显示
 *   expired    = 已解决超过 30 天无 reopen, 不显示
 */
export type IssueStatus = 'unresolved' | 'reopened' | 'resolved' | 'expired'

export interface TelemetryEventRow {
  id: number
  ts: number
  serverTs: string
  type: string
  category: string
  action: string
  label: string
  value: number
  path: string
  userId: number
  sessionId: string
  platform: string
  appVersion: string
  extra: string
  resolution?: IssueResolution
  issueStatus: IssueStatus
}

export interface ListEventsResp {
  total: number
  rows: TelemetryEventRow[]
}

export interface FilmStat {
  mid: number
  name: string
  cover: string
  count: number
}

export interface ApiPerfItem {
  action: string
  count: number
  avg: number
  p50: number
  p95: number
  p99: number
}

export const overview = (days = 7): Promise<OverviewResp> =>
  http.get('/manage/telemetry/overview', { params: { days } })

export const listEvents = (params: {
  type?: string
  days?: number
  limit?: number
  offset?: number
  /** active=未解决+重启+宽限期(默认) | all | resolved | reopened */
  status?: 'active' | 'all' | 'resolved' | 'reopened'
}): Promise<ListEventsResp> =>
  http.get('/manage/telemetry/events', { params })

/** 标记某 issue 已解决 — key = (category,path,label) 必须与 telemetry 行一致 */
export const resolveIssue = (body: {
  category: string
  path: string
  label: string
  commitId: string
  note?: string
}): Promise<void> => http.post('/manage/telemetry/resolve', body)

/** 撤销解决 (重新标记为未解决) */
export const unresolveIssue = (body: {
  category: string
  path: string
  label: string
}): Promise<void> => http.post('/manage/telemetry/unresolve', body)

/** 热点视频 — 按访问量 (PV) 排序. hours = days * 24, 服务端默认 168h(7天) */
export const hotFilms = (days = 7, limit = 10): Promise<FilmStat[]> =>
  http.get('/manage/telemetry/hot-films', { params: { hours: days * 24, limit } })

/** 收藏最多 — 按收藏数排序 */
export const mostFavorited = (limit = 10): Promise<FilmStat[]> =>
  http.get('/manage/telemetry/most-favorited', { params: { limit } })

export const apiPerf = (days = 7, limit = 10): Promise<ApiPerfItem[]> =>
  http.get('/manage/telemetry/api-perf', { params: { days, limit } })

File: server/internal/handler/telemetry.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"server/internal/dto"
)

// TelemetryOverview GET /manage/telemetry/overview?days=
func (h *Handlers) TelemetryOverview(c *gin.Context) {
	ov, err := h.Telemetry.Overview(c.Request.Context(), queryInt(c, "days", 7))
	respond(c, ov, err)
}

// TelemetryEvents GET /manage/telemetry/events?type=&days=&status=&limit=&offset=
// 看板错误列表: 前端按 type=error 拉, 含问题闭环 issueStatus, 返回 {rows,total}。
func (h *Handlers) TelemetryEvents(c *gin.Context) {
	rows, total, err := h.Telemetry.ErrorEvents(
		c.Request.Context(),
		queryInt(c, "days", 7),
		c.Query("type"),
		c.DefaultQuery("status", "active"),
		queryInt(c, "limit", 100),
		queryInt(c, "offset", 0),
	)
	if err != nil {
		dto.Fail(c, err)
		return
	}
	dto.OK(c, gin.H{"rows": rows, "total": total})
}

// TelemetryPurge POST /manage/telemetry/purge-bots — 清理历史脏数据(bot/空UA/垃圾路径)。
func (h *Handlers) TelemetryPurge(c *gin.Context) {
	n, err := h.Telemetry.PurgeNoise(c.Request.Context())
	if err != nil {
		dto.Fail(c, err)
		return
	}
	dto.OK(c, gin.H{"deleted": n})
}

// TelemetryPurgeApiErrors POST /manage/telemetry/purge-api-errors —
// 一次性清理误记的 API HTTP 错误历史(extra.type=error 且 category 以 api- 前缀, 如 api-4xx/api-5xx)。
func (h *Handlers) TelemetryPurgeApiErrors(c *gin.Context) {
	n, err := h.Telemetry.PurgeApiErrors(c.Request.Context())
	if err != nil {
		dto.Fail(c, err)
		return
	}
	dto.OK(c, gin.H{"deleted": n})
}

// TelemetryPurgeErrorsBefore POST /manage/telemetry/purge-errors-before?days=5 —
// 按龄清理: 删除 days 天前的错误事件(extra.type=error 且 server_ts < 截止), 默认 5 天。
func (h *Handlers) TelemetryPurgeErrorsBefore(c *gin.Context) {
	n, err := h.Telemetry.PurgeErrorsBefore(c.Request.Context(), queryInt(c, "days", 5))
	if err != nil {
		dto.Fail(c, err)
		return
	}
	dto.OK(c, gin.H{"deleted": n})
}

// TelemetryPurgeChunkNoise POST /manage/telemetry/purge-chunk-noise —
// 清理 chunk 加载失败类噪声错误(部署后旧 hash 自愈现象, 非真异常)。
func (h *Handlers) TelemetryPurgeChunkNoise(c *gin.Context) {
	n, err := h.Telemetry.PurgeChunkNoise(c.Request.Context())
	if err != nil {
		dto.Fail(c, err)
		return
	}
	dto.OK(c, gin.H{"deleted": n})
}

// TelemetryTopPaths GET /manage/telemetry/top-paths?hours=&limit=
func (h *Handlers) TelemetryTopPaths(c *gin.Context) {
	list, err := h.Telemetry.TopPaths(c.Request.Context(), queryInt(c, "hours", 24), queryInt(c, "limit", 20))
	respond(c, list, err)
}

// TelemetryApiPerf GET /manage/telemetry/api-perf?days=&limit=
func (h *Handlers) TelemetryApiPerf(c *gin.Context) {
	list, err := h.Telemetry.ApiPerf(c.Request.Context(), queryInt(c, "days", 7), queryInt(c, "limit", 10))
	respond(c, list, err)
}

// TelemetryHotFilms GET /manage/telemetry/hot-films?hours=&limit=  (热点视频: 按埋点 pv 访问量)
func (h *Handlers) TelemetryHotFilms(c *gin.Context) {
	list, err := h.Telemetry.HotFilms(c.Request.Context(), queryInt(c, "hours", 24*7), queryInt(c, "limit", 10))
	respond(c, list, err)
}

// TelemetryMostFavorited GET /manage/telemetry/most-favorited?limit=  (收藏最多)
func (h *Handlers) TelemetryMostFavorited(c *gin.Context) {
	list, err := h.Telemetry.MostFavorited(c.Request.Context(), queryInt(c, "limit", 10))
	respond(c, list, err)
}

type issueReq struct {
	Category string `json:"category"`
	Path     string `json:"path"`
	Label    string `json:"label"`
	CommitId string `json:"commitId"`
}

// TelemetryResolve POST /manage/telemetry/resolve
func (h *Handlers) TelemetryResolve(c *gin.Context) {
	var req issueReq
	if err := c.ShouldBindJSON(&req); err != nil {
		dto.Error(c, http.StatusUnprocessableEntity, "invalid body")
		return
	}
	if err := h.Telemetry.ResolveIssue(c.Request.Context(), req.Category, req.Path, req.Label, req.CommitId); err != nil {
		dto.Fail(c, err)
		return
	}
	dto.OK(c, gin.H{"ok": true})
}

// TelemetryUnresolve POST /manage/telemetry/unresolve
func (h *Handlers) TelemetryUnresolve(c *gin.Context) {
	var req issueReq
	if err := c.ShouldBindJSON(&req); err != nil {
		dto.Error(c, http.StatusUnprocessableEntity, "invalid body")
		return
	}
	if err := h.Telemetry.UnresolveIssue(c.Request.Context(), req.Category, req.Path, req.Label); err != nil {
		dto.Fail(c, err)
		return
	}
	dto.OK(c, gin.H{"ok": true})
}
